pkg/util/tracingservice/client_test.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tracingservice
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/migration/nodelivenesstest"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/rpc"
+	"github.com/cockroachdb/cockroach/pkg/rpc/nodedialer"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/netutil"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/cockroach/pkg/util/tracingservice/tracingservicepb"
+	"github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func setUpService(
+	t testing.TB,
+	rpcContext *rpc.Context,
+	localNodeID roachpb.NodeID,
+	remoteNodeID roachpb.NodeID,
+	localTracer *tracing.Tracer,
+	remoteTracer *tracing.Tracer,
+) *TraceClientDialer {
+	s := rpc.NewServer(rpcContext)
+	remoteTraceServer := NewTraceService(remoteTracer)
+	tracingservicepb.RegisterTraceServer(s, remoteTraceServer)
+	ln, err := netutil.ListenAndServeGRPC(rpcContext.Stopper, s, util.TestAddr)
+	require.NoError(t, err)
+
+	s2 := rpc.NewServer(rpcContext)
+	localBlobServer := NewTraceService(localTracer)
+	require.NoError(t, err)
+	tracingservicepb.RegisterTraceServer(s2, localBlobServer)
+	ln2, err := netutil.ListenAndServeGRPC(rpcContext.Stopper, s2, util.TestAddr)
+	require.NoError(t, err)
+
+	localDialer := nodedialer.New(rpcContext,
+		func(nodeID roachpb.NodeID) (net.Addr, error) {
+			if nodeID == remoteNodeID {
+				return ln.Addr(), nil
+			} else if nodeID == localNodeID {
+				return ln2.Addr(), nil
+			}
+			return nil, errors.Errorf("node %d not found", nodeID)
+		},
+	)
+
+	testNodeLiveness := nodelivenesstest.New(2 /* numNodes */)
+	return NewTraceClientDialer(
+		localNodeID,
+		localDialer,
+		testNodeLiveness,
+		localTracer,
+	)
+}
+
+func TestTraceClientGetSpanRecordings(t *testing.T) {
+	localNodeID := roachpb.NodeID(1)
```

Comment:
Have you considered using testcluster instead? Instead of using "raw tracers"? I think that'd be the better thing to do for these fake-distributed tests, cause you're interesting in making sure that the "wiring" is all correct (i.e. we find the right set of nodes, etc.), and that the right service endpoints are registered with grpc/etc. Right now it feels like it wants to be an integration test, but is setting up all the integration by hand.

Reply:
yeah good point, changed to using a TC with 2 nodes. Unfortunately, we can't have` tracingservice` depend on `testcluster` or `serverutils` because we get into a:

`tracingservice` -> `testcluster` -> `server` -> `tracingservice`

So I had to move the tests into their own subpackage.

Reply:
What's customary in these cases is using `package tracingservice_test` for the test files, and keeping them in the same package. Is that possible here?